plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/EqualityDeleteFilter.java

```diff
@@ -17,51 +17,227 @@
 import io.trino.spi.Page;
 import io.trino.spi.connector.ConnectorPageSource;
 import io.trino.spi.type.Type;
-import org.apache.iceberg.StructLike;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructLikeMap;
+import org.apache.iceberg.util.StructProjection;
 
+import java.util.HashMap;
 import java.util.List;
-import java.util.function.Consumer;
-import java.util.function.Predicate;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
+import static io.trino.plugin.iceberg.IcebergUtil.schemaFromHandles;
 import static java.util.Objects.requireNonNull;
 
 public final class EqualityDeleteFilter
         implements DeleteFilter
 {
-    private final Predicate<StructLike> deletedRows;
+    /**
+     * Box class for data sequence number to avoid auto-boxing when adding values to the delete map
+     */
+    private static class DataSequenceNumber
+    {
+        private long dataSequenceNumber;
+
+        public DataSequenceNumber(long dataSequenceNumber)
+        {
+            this.dataSequenceNumber = dataSequenceNumber;
+        }
+    }
 
-    public EqualityDeleteFilter(Predicate<StructLike> deletedRows)
+    public static final class FileToken
     {
-        this.deletedRows = requireNonNull(deletedRows, "deletedRows is null");
+        private long dataVersion;
+        private boolean isLoaded;
+        private boolean isLoading;
+        private final Object lockToken;
```

Comment:
I'm not sure about the synchronization semantics here, it seems like you could use `ListenableFuture` to denote currently-loading or already loaded values and not manage the state via `isLoading` / `isLoaded` in here

Reply:
I have not tackled this yet, if this is something you are still interested in let me know. The current load mechanism uses the split's thread directly. This can be achieved with a "Future" that runs on the local thread.